File: src/main/java/darkorg/betterdurability/util/VanillaDamageableType.java
package darkorg.betterdurability.util;

import net.minecraft.world.entity.EquipmentSlot;
import net.minecraft.world.item.*;

import javax.annotation.Nullable;
import java.util.Arrays;

import static darkorg.betterdurability.event.DurabilityEventsHandler.isWhitelisted;

public enum VanillaDamageableType {
    // brokenThreshold and protectValue here is somewhat arbitrary
    AXE(Category.TOOL, 2, 1) {
        @Override
        public boolean isItemThisType(Item item) {
            return item instanceof AxeItem;
        }
    },
    PICKAXE(Category.TOOL, 2, 1) {
        @Override
        public boolean isItemThisType(Item item) {
            return item instanceof PickaxeItem;
        }
    },
    SHOVEL(Category.TOOL, 2, 1) {
        @Override
        public boolean isItemThisType(Item item) {
            return item instanceof ShovelItem;
        }
    },
    HOE(Category.TOOL, 1, 1) {
        @Override
        public boolean isItemThisType(Item item) {
            return item instanceof HoeItem;
        }
    },
    SHEARS(Category.TOOL, 1, 1) {
        @Override
        public boolean isItemThisType(Item item) {
            return item instanceof ShearsItem;
        }
    },
    SWORD(Category.TOOL, 2,1) {
        @Override
        public boolean isItemThisType(Item item) {
            return item instanceof SwordItem;
        }
    },
    FISHING_ROD(Category.TOOL, 5, 2) {
        @Override
        public boolean isItemThisType(Item item) {
            return item instanceof FishingRodItem;
        }
    },
    FLINT_AND_STEEL(Category.TOOL, 1, 1) {
        @Override
        public boolean isItemThisType(Item item) {
            return item instanceof FlintAndSteelItem;
        }
    },
    BOW(Category.TOOL, 1,1) {
        @Override
        public boolean isItemThisType(Item item) {
            return item instanceof BowItem;
        }
    },
    TRIDENT(Category.TOOL, 2, 1) {
        @Override
        public boolean isItemThisType(Item item) {
            return item instanceof TridentItem;
        }
    },
    CROSSBOW(Category.TOOL, 9, 4) {
        @Override
        public boolean isItemThisType(Item item) {
            return item instanceof CrossbowItem;
        }
    },
    HELMET(Category.ARMOR, 2, 1) {
        public boolean isItemThisType(Item item) {
            return item instanceof ArmorItem armorItem && armorItem.getEquipmentSlot() == EquipmentSlot.HEAD;
        }
    },
    CHESTPLATE(Category.ARMOR, 2, 1) {
        @Override
        public boolean isItemThisType(Item item) {
            return item instanceof ArmorItem armorItem && armorItem.getEquipmentSlot() == EquipmentSlot.CHEST;
        }
    },
    LEGGINGS(Category.ARMOR, 2, 1) {
        @Override
        public boolean isItemThisType(Item item) {
            return item instanceof ArmorItem armorItem && armorItem.getEquipmentSlot() == EquipmentSlot.LEGS;
        }
    },
    BOOTS(Category.ARMOR, 2, 1) {
        @Override
        public boolean isItemThisType(Item item) {
            return item instanceof ArmorItem armorItem && armorItem.getEquipmentSlot() == EquipmentSlot.FEET;
        }
    },
    SHIELD(Category.SHIELD, 4, 2) {
        @Override
        public boolean isItemThisType(Item item) {
            return item instanceof ShieldItem;
        }
    };

    public enum Category { TOOL, ARMOR, SHIELD }

    public final Category category;
    public final int brokenThreshold; // damageable things will lose functionality if durability not more than this
    public final int protectValue; // durability will never drop below this, should be not more than brokenThreshold

    VanillaDamageableType(Category category, int brokenThreshold, int protectValue) {
        assert protectValue > 0 && brokenThreshold >= protectValue;
        this.category = category;
        this.brokenThreshold = brokenThreshold;
        this.protectValue = protectValue;
    }

    public boolean isItemBroken(ItemStack stack) {
        int durabilityLeft = stack.getMaxDamage() - stack.getDamageValue();
        return durabilityLeft <= this.brokenThreshold;
    }

    public abstract boolean isItemThisType(Item item);

    @Nullable
    public static VanillaDamageableType getTypeByItem(Item item) {
        return Arrays.stream(VanillaDamageableType.values())
                .filter(t -> t.isItemThisType(item))
                .findAny().orElse(null);
    }

    public static boolean isItemKnownBroken(ItemStack stack) {
        if (stack.isDamageableItem()) {
            VanillaDamageableType itemType = VanillaDamageableType.getTypeByItem(stack.getItem());
            return itemType != null && itemType.isItemBroken(stack);
        }
        return false;
    }

    public static boolean isItemKnownBrokenAnother (ItemStack stack) {
        if (stack.isDamageableItem()) {
            int durabilityLeft = stack.getMaxDamage() - stack.getDamageValue();
            return isWhitelisted(stack.getItem()) && durabilityLeft <= 2;
        }
        return false;
    }




}

File: forge/shared/src/main/kotlin/darkorg/betterdurability/util/VanillaDamageableType.kt
package darkorg.betterdurability.util

import darkorg.betterdurability.polyfill.minecraft.inventory.EquipmentType
import darkorg.betterdurability.polyfill.minecraft.item.*


typealias VanillaDamageableCategory = VanillaDamageableType.Category

enum class VanillaDamageableType(category: Category, brokenThreshold: Int, protectValue: Int) {
    // brokenThreshold and protectValue here is somewhat arbitrary
    AXE(Category.TOOL, 2, 1) {
        override fun isItemThisType(item: Item): Boolean {
            return item is AxeItem
        }
    },
    PICKAXE(Category.TOOL, 2, 1) {
        override fun isItemThisType(item: Item): Boolean {
            return item is PickaxeItem
        }
    },
    SHOVEL(Category.TOOL, 2, 1) {
        override fun isItemThisType(item: Item): Boolean {
            return item is ShovelItem
        }
    },
    HOE(Category.TOOL, 1, 1) {
        override fun isItemThisType(item: Item): Boolean {
            return item is HoeItem
        }
    },
    SHEARS(Category.TOOL, 1, 1) {
        override fun isItemThisType(item: Item): Boolean {
            return item is ShearsItem
        }
    },
    SWORD(Category.TOOL, 2, 1) {
        override fun isItemThisType(item: Item): Boolean {
            return item is SwordItem
        }
    },
    FISHING_ROD(Category.TOOL, 5, 2) {
        override fun isItemThisType(item: Item): Boolean {
            return item is FishingRodItem
        }
    },
    FLINT_AND_STEEL(Category.TOOL, 1, 1) {
        override fun isItemThisType(item: Item): Boolean {
            return item is FlintAndSteelItem
        }
    },
    BOW(Category.TOOL, 1, 1) {
        override fun isItemThisType(item: Item): Boolean {
            return item is BowItem
        }
    },
    TRIDENT(Category.TOOL, 2, 1) {
        override fun isItemThisType(item: Item): Boolean {
            return item is TridentItem
        }
    },
    CROSSBOW(Category.TOOL, 9, 4) {
        override fun isItemThisType(item: Item): Boolean {
            return item is CrossbowItem
        }
    },
    HELMET(Category.ARMOR, 2, 1) {
        override fun isItemThisType(item: Item): Boolean {
            return item is ArmorItem && item.slot == EquipmentType.HEAD
        }
    },
    CHESTPLATE(Category.ARMOR, 2, 1) {
        override fun isItemThisType(item: Item): Boolean {
            return item is ArmorItem && item.slot == EquipmentType.CHEST
        }
    },
    LEGGINGS(Category.ARMOR, 2, 1) {
        override fun isItemThisType(item: Item): Boolean {
            return item is ArmorItem && item.slot == EquipmentType.LEGS
        }
    },
    BOOTS(Category.ARMOR, 2, 1) {
        override fun isItemThisType(item: Item): Boolean {
            return item is ArmorItem && item.slot == EquipmentType.FEET
        }
    },
    SHIELD(Category.SHIELD, 4, 2) {
        override fun isItemThisType(item: Item): Boolean {
            return item is ShieldItem
        }
    };

    enum class Category { TOOL, ARMOR, SHIELD }

    @JvmField val category: Category
    @JvmField val brokenThreshold: Int // damageable things will lose functionality if durability not more than this
    @JvmField val protectValue: Int // durability will never drop below this, should be not more than brokenThreshold

    init {
        assert(protectValue in 1..brokenThreshold)
        this.category = category
        this.brokenThreshold = brokenThreshold
        this.protectValue = protectValue
    }

    fun isItemBroken(stack: ItemStack): Boolean {
        val durabilityLeft = stack.maxDamage - stack.damageValue
        return durabilityLeft <= this.brokenThreshold
    }

    abstract fun isItemThisType(item: Item): Boolean

    companion object {
        @JvmStatic fun getTypeByItem(item: Item): VanillaDamageableType? {
            return values().firstOrNull { t -> t.isItemThisType(item) }
        }

        @JvmStatic fun isItemKnownBroken(stack: ItemStack): Boolean {
            if (stack.isDamageableItem) {
                val itemType = getTypeByItem(stack.item)
                return itemType != null && itemType.isItemBroken(stack)
            }
            return false
        }
    }
}
